### DISCUSSION TITLE
Question: How do you use getDecodedData?

### DISCUSSION THREAD

**Discussion by doctor30309** [katspaugh/wavesurfer.js#2932] [Q&A] [answered]
I recently converted tot he beta version.  I was using exportPCM to get the peak data.  I converted to getDecodedData as instructed.  But I don't understand how to use it now..

Here is the js code: wavesurfer.getDecodedData().then( send json data to service api to save on disc);

I am getting an error "wavesurfer.getDecodedData(...).then is not a function".  This worked with exportPCM.

Could use some guidance on what I need to change with the Beta version to get this to work...

Thanks in advance!

**Comment by katspaugh** [comment]
It’s a synchronous function, just assign it to a variable like so:
```js
const data = wavesurfer.getDecodedData()
```
Keep in mind it’s an [AudioBuffer](https://developer.mozilla.org/en-US/docs/Web/API/AudioBuffer) that is returned. So to get the peaks:
```js
const peaks = data.getChannelData(0)
```

**Reply by doctor30309** [reply]
Thanks!  I did find my way into the documentation on AudioBuffer and resting code now.

**Comment by doctor30309** [comment]
@katspaugh  - So I have been messing around to get this to work for a couple of days.

Here is the code snippet..  However, when the post occurs, the value of the parameter MyJson is null.  So something is happening with the json string being passed.  I can pass the word "test" just fine...    

You'll notice I write the data to the log.  This looks like a valid JSON object.

This code worked fine when exportPCM function was in the library.  What am I missing??

    var data = wavesurfer.getDecodedData().getChannelData(0);
    console.log(JSON.stringify(pcmArray));

    $.ajax({
        type: "POST",
        dataType: "json",
        url: 'Admin/SavePeakFile/',
        data: { myJson: JSON.stringify(data)},
        done: function (result) {
            console.log("Peak File saved successfully.");
        },
        fail: function (result) {
            console.log("Error saving Peak File: " + result.value);
        }
    });

**Reply by katspaugh** [reply]
Your code looks correct except for the `console.log(JSON.stringify(pcmArray))` part. `pcmArray` is not defined anywhere in the snippet.

**Reply by doctor30309** [reply]
Hi.

That was a typo as I was cleaning up the routine to show you the basics...  I did make one change as the original code was also returning the Array index as part of the json string.  I used the Array.from to strip that out.

However, I can't seem to get the binding back to my MVC controller to work when I issue the POST.  variable names (myJson and songName are exact matches in the SavePeakData routine...

Question: in the exportPCM function that was deprecated, was there any special encoding you did?  I know it's a stretch but I am grasping at straws right now...

   var data = wavesurfer.getDecodedData().getChannelData(0);
    console.log(JSON.stringify(Array.from(data)));

    $.ajax({
        type: "POST",
        dataType: "json",
        url: 'Admin/SavePeakFile/',
        data: { myJson: JSON.stringify(Array.from(data)), songName: songName },
        done: function (result) {
            console.log("Peak File saved successfully.");
        },
        fail: function (result) {
            console.log("Error saving Peak File: " + result.value);
        }
    });

**Reply by doctor30309** [reply]
Here is the controller code.  Nothing has changed here since I was using exportPCM successfully...  The problem again is that both parameters (myJson and songName) are null when I hit my breakpoint to debug...

        [Route("Admin/SavePeakFile/")]
        [HttpPost]
        public async Task<IActionResult> SavePeakFileAsync(string myJson, string songName)
        {
            string myMessage = (myJson != null) ? "Not Null" : "isNull";

            var n = JsonSerializer.Deserialize<float[]>(myJson!);

            string PeakFileLocation = Path.Combine(_env.WebRootPath, "Music/PeakFiles");
            string PeakFileName = Path.Combine(PeakFileLocation, songName + ".json");

            if (!Directory.Exists(PeakFileLocation)) { Directory.CreateDirectory(PeakFileLocation); }

            if (!System.IO.File.Exists(PeakFileName))
            {
                await using FileStream createStream = System.IO.File.Create(PeakFileName);
                await JsonSerializer.SerializeAsync(createStream, myJson);
            }

            return Json(myMessage);
        }

**Reply by katspaugh** [reply]
The old [exportPCM did some mapping](https://github.com/katspaugh/wavesurfer.js/blob/v6/src/wavesurfer.js#L1660-L1695) but in the end it returned an array of floats of arbitrary precision and given length. Your controller doesn't seem to care about either, so it should still work as long as you send the JSON string correctly.

One thing that just occurred to me is that there's likely a double JSON encoding going on. jQuery's data field is already JSON-encoded by jQuery internally, so you don't need to stringify on top of that. Just pass a raw array:
```
data: { myJson: Array.from(data), songName: songName },
``` 
Hope this helps.

**Reply by doctor30309** [reply]
I'll give that a try.

Before I head down this road to figure out how to re-create your export{CM function, is there any plans to bring back the exportPCM function to the library?

**Reply by katspaugh** [reply]
Feel free to create a PR for it.

I can imagine it should be something like:
```js
exportPeaks: (): Array<Float32Array | number[]> => {
  const buffer = wavesurfer.getDecodedData()
  const peaks = [ buffer.getChannelData(0) ]
  if (buffer.numberOfChannels > 1) {
    peaks.push(buffer.getChannelData(1))
  }
  return peaks
}
```

As you can see it would still be quite different from the old one:
 * returns an array, not a promise
 * in fact, it would return an array of arrays of floats (the first one or two channels)

**Reply by robots4life** [reply]
@katspaugh I also recently started using version 7 and have some questions on how to properly use it. The new site looks nice, the examples are cool to explore, the docs are also there, however I am still missing some pieces. Thank you to you and all the contributes that put so much work into this. :+1: 

1.
```js
		wavesurfer = WaveSurfer.create({
			container: container,
			waveColor: 'green',
			progressColor: 'yellow',
			cursorWidth: 2,
			cursorColor: 'red',
			responsive: true
			// url: audio
		});

		wavesurfer.load(audio);
```
Do I need to use the "url" key to pass the audio to WaveSurfer or can I use the `load` method ?

2. on the peaks..
```js
		wavesurfer.on('ready', () => {
			console.log('ready');

			const buffer = wavesurfer.getDecodedData();
			// console.log(buffer);

			let channel0Peaks = buffer.getChannelData(0);
			console.log('channel0Peaks : ', channel0Peaks);

			let channel1Peaks = buffer.getChannelData(1);
			console.log('channel1Peaks : ', channel1Peaks);

			console.log('buffer length : ', buffer.length);
			console.log('buffer duration : ', buffer.duration);
			console.log('buffer sampleRate : ', buffer.sampleRate);
			console.log('buffer numberOfChannels : ', buffer.numberOfChannels);
		});
```
What is the correct event to use in order to get the channel data ? `ready`, `load`, `decode` or `interaction` ?

3. How can I pass the generated peaks to WaveSurfer ?

Ok from here https://wavesurfer-js.org/docs/classes/wavesurfer.WaveSurfer#load I can see it accepts a Float32Array, does that mean I can just pass it like I get it from getChannelData ?

But how do I pass the data to the `load` method when I have more than 1 channel ? Do I need to combine the data from both channels I get back from each channel ? How do I combine it into something that the `load` method will accept ? Like this ?
```js
const peaks = [ [channel_1_DataFloat32Array], [channel_2_DataFloat32Array] ];
```
A single Float32Array has 1127025 items, I also wonder if there is some optimization that can be used to store the data in a database ? I think this might cause some issues when sending the the data to the database. Is there something that can be done about it ? Like the `exportPCM` method had 4 parameters.
```js
wavesurfer.exportPCM(1024, 10000, true, 0)
```
I think the 1024 stood for the array items to be stored, no ? Is there something similar I can do with the Float32Array to be able to better handle the storing and fetching of this PCM data on page load ?

Long story short, could you please be so kind as to provide an example of how `getChannelData` can be used on each channel of the buffer and then passed back to the `load` method ?

Thank you very much for any help with this. :smile: :+1:

**Reply by katspaugh** [reply, marked as answer]
Appreciate the feedback!

## 1. Load, URL and peaks
Regarding the `url` and `peaks` properties, both can be passed as either a create option or as arguments to load.

As `create` options:
```js
const wavesurfer = WaveSurfer.create({
  container: '#container',
  url: 'https://...',
  peaks: [[...], [...]]
})
```

As `load` arguments:
```js
const wavesurfer = WaveSurfer.create({
  container: '#container',
})

wavesurfer.load('https://...', [[...], [...]])
```

## 2. Exporting peaks

Yes, correct, you can pass `channelData` arrays directly as peaks, one array per channel.

E.g.
```ts
const exportPeaks = (): Array<Float32Array> => {
  const buffer = wavesurfer.getDecodedData()
  const peaks = []
  for (let i = 0; i < buffer.numberOfChannels; i++) {
    peaks.push(buffer.getChannelData(i))
  }
  return peaks
}
```

The best event for that is `decode`, it will be fired as soon as it's decoded but before it's rendered:

```js
wavesurfer.on('decode', () => {
  const peaks = exportPeaks()
  saveToDb(peaks)
})
```

I should probably restore `exportPCM` indeed, as it seems the new method isn't super straight-forward to use.
Edit: added an issue for it: #3007.
